server/src/main/java/com/google/android/attestation/AuthorizationList.java

```diff
@@ -314,4 +361,197 @@ static Set<UserAuthType> userAuthTypeToEnum(long userAuthType) {
+        KM_TAG_ATTESTATION_ID_MANUFACTURER, this.attestationIdManufacturer, vector);
+    addOptionalOctetString(KM_TAG_ATTESTATION_ID_MODEL, this.attestationIdModel, vector);
+    addOptionalInteger(KM_TAG_VENDOR_PATCH_LEVEL, this.vendorPatchLevel, vector);
+    addOptionalInteger(KM_TAG_BOOT_PATCH_LEVEL, this.bootPatchLevel, vector);
+    addBoolean(KM_TAG_DEVICE_UNIQUE_ATTESTATION, this.individualAttestation, vector);
+    return new DERSequence(vector);
+  }
+
+  private static void addOptionalIntegerSet(
+      int tag, Optional<Set<Integer>> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      ASN1EncodableVector tmp = new ASN1EncodableVector();
+      entry.get().forEach((Integer value) -> tmp.add(new ASN1Integer(value.longValue())));
+      vector.add(new DERTaggedObject(tag, new DERSet(tmp)));
+    }
+  }
+
+  private static void addOptionalInstant(
+      int tag, Optional<Instant> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new ASN1Integer(entry.get().toEpochMilli())));
+    }
+  }
+
+  private static void addOptionalDuration(
+      int tag, Optional<Duration> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new ASN1Integer(entry.get().getSeconds())));
+    }
+  }
+
+  private static void addBoolean(int tag, boolean entry, ASN1EncodableVector vector) {
+    if (entry) {
+      vector.add(new DERTaggedObject(tag, DERNull.INSTANCE));
+    }
+  }
+
+  private static void addOptionalInteger(
+      int tag, Optional<Integer> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new ASN1Integer(entry.get())));
+    }
+  }
+
+  private static void addOptionalLong(int tag, Optional<Long> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new ASN1Integer(entry.get())));
+    }
+  }
+
+  private static void addOptionalOctetString(
+      int tag, Optional<byte[]> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new DEROctetString(entry.get())));
+    }
+  }
+
+  private static void addOptionalUserAuthType(
+      int tag, Optional<Set<UserAuthType>> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, new ASN1Integer(userAuthTypeToLong(entry.get()))));
+    }
+  }
+
+  private static void addOptionalRootOfTrust(
+      int tag, Optional<RootOfTrust> entry, ASN1EncodableVector vector) {
+    if (entry.isPresent()) {
+      vector.add(new DERTaggedObject(tag, entry.get().toAsn1Sequence()));
+    }
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Builder for an AuthorizationList. Any field not set will be made an Optional.empty or set with
+   * the default value.
+   */
+  public static final class Builder {
```

Comment:
Would having setter methods make using this builder easier?

Reply:
Done.